### PR TITLE
PB-1831: Kopia restore changes.

### PR DESCRIPTION
**Pull request by diptiranjanpx**
Includes driver, executor, reconciler, and commands changes.

Signed-off-by: Diptiranjan

**What this PR does / why we need it**:
This PR has support for restoring backups with Kopia.
 - Added new driver kopiarestore
 - Restore changes as part of the executor
 - Getting the restore command and the run implementation
 - Reconciler changes while the driver is kopia restore.

**Test**:

Following tests have been done manually.
1. Deploying MySQL, writing data, taking a kopia backup, restoring to the same namespace but creating a new deployment, data integrity check.
2. Deploying MySQL, writing data, taking a kopia backup, restoring to a different namespace, and doing data integrity check.

All test details have been updated in PB-1831.

### CODE REVIEW

File: pkg/drivers/kopiarestore/kopiarestore.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+		return nil, err
+	}
+
+	cmd := strings.Join([]string{
+		"/kopiaexecutor",
+		"restore",
+		"--backup-location",
+		backuplocationName,
+		"--namespace",
+		backuplocationNamespace,
```
Comment (prashanthpx) [inline pkg/drivers/kopiarestore/kopiarestore.go:151]: This should be namespace (even though BL ns and these are the same) for readability.

File: pkg/kopia/restore.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	if snapshotID == "" {
+		return nil, fmt.Errorf("snapshot id cannot be empty")
+	}
+
+	args := []string{snapshotID, "."}
+
+	return &Command{
+		Name:     "restore",
+		Password: password,
+		Dir:      DefaultDestPath,
```
Comment (prashanthpx) [inline pkg/kopia/restore.go:39]: Shouldn't this be assigned with `targetPath` as this is passed from job caller

File: pkg/kopia/restore.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		errBuf: new(bytes.Buffer),
+	}
+}
+
+func (b *restoreExecutor) Run() error {
+
+	b.responseLock.Lock()
+	defer b.responseLock.Unlock()
+
+	if b.isRunning {
```
Comment (prashanthpx) [inline pkg/kopia/restore.go:84]: We can remove this as this is not likely to happen (caller calling the same instance of run again with same executor instance)

File: pkg/kopia/restore.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+	if err := b.execCmd.Start(); err != nil {
+		b.lastError = err
+		return err
+	}
+	b.isRunning = true
+
+	go func() {
+		err := b.execCmd.Wait()
+		b.responseLock.Lock()
```
Comment (prashanthpx) [inline pkg/kopia/restore.go:100]: Locks are not used anymore, can be removed.

**Review by siva-portworx — APPROVED** [review]

File: hack/tools.go [hunk 5]
```diff
@@ -1,3 +1,4 @@
+//go:build tools
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply hack/tools.go:1]: Running make check-fmt added these.

File: pkg/controllers/dataexport/reconcile.go [hunk 6]
```diff
@@ -135,6 +135,37 @@ func (c *Controller) sync(ctx context.Context, in *kdmpapi.DataExport) (bool, er
...
+				return false, c.updateStatus(dataExport, kdmpapi.DataExportStatusFailed, msg)
+			}
+			// This will create a unique secret per PVC being restored
+			err = CreateCredentialsSecret(
+				utils.FrameCredSecretName(dataExport.Name, dataExport.Spec.Destination.Name),
+				vb.Spec.BackupLocation.Name,
+				vb.Spec.BackupLocation.Namespace,
+			)
+			if err != nil {
+				msg := fmt.Sprintf("failed to create cloud credential secret: %v", err)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/controllers/dataexport/reconcile.go:161]: Done

File: pkg/controllers/dataexport/reconcile.go [hunk 7]
```diff
@@ -470,6 +501,16 @@ func startTransferJob(drv drivers.Interface, srcPVCName string, dataExport *kdmp
 			drivers.WithLabels(jobLabels(dataExport.GetName())),
 			drivers.WithDataExportName(dataExport.GetName()),
 		)
+	case drivers.KopiaRestore:
+		return drv.StartJob(
+			drivers.WithSourcePVC(srcPVCName),
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/controllers/dataexport/reconcile.go:506]: Was using it to create the secret name, but I think it is better to use the destination PVC name to frame the secret name. have removed this.

File: pkg/executor/kopia/restore.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+func newRestoreCommand() *cobra.Command {
+	var (
+		targetPath string
+		snapshotID string
+	)
+	restoreCommand := &cobra.Command{
+		Use:   "restore",
+		Short: "Start a kopia restore",
+		Run: func(c *cobra.Command, args []string) {
+			if len(backupLocationFile) == 0 && len(backupLocationName) == 0 {
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/executor/kopia/restore.go:23]: Removed.

File: pkg/executor/kopia/restore.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+				util.CheckErr(fmt.Errorf("backup-location or backup-location-file has to be provided for kopia backups"))
+				return
+			}
+			if len(targetPath) == 0 {
+				util.CheckErr(fmt.Errorf("target-path argument is required for kopia backups"))
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/executor/kopia/restore.go:28]: done

File: pkg/kopia/restore.go (hunk 2 above)
Comment (diptiranjanpx) [reply pkg/kopia/restore.go:39]: yes, done

File: pkg/kopia/restore.go (hunk 3 above)
Comment (diptiranjanpx) [reply pkg/kopia/restore.go:84]: removed.

File: pkg/kopia/restore.go (hunk 4 above)
Comment (diptiranjanpx) [reply pkg/kopia/restore.go:100]: done

File: pkg/executor/kopia/restore.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	restoreCommand.Flags().StringVar(&snapshotID, "id", "", "Snapshot id of the backup")
+	return restoreCommand
+}
+
+func runRestore(snapshotID, targetPath string) error {
+	logrus.Infof("Restore started from snapshotID: %s", snapshotID)
+	// Parse using the mounted secrets
+	fn := "runRestore"
+	repo, err := executor.ParseCloudCred()
+	repo.Name = kopiaRepo
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/executor/kopia/restore.go:45]: done

File: pkg/executor/kopia/restore.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+			volumeBackupName,
+			namespace,
+		); statusErr != nil {
+			return statusErr
+		}
+		return fmt.Errorf("parse backuplocation: %s", err)
+	}
+
+	if err = runKopiaRepositoryConnect(repo); err != nil {
+		errMsg := fmt.Sprintf("repository connect failed: %v", err)
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/executor/kopia/restore.go:59]: done

File: pkg/drivers/kopiarestore/kopiarestore.go [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+	return utils.ToJobStatus(drivers.TransferProgressCompleted, ""), nil
+}
+
+func (d Driver) validate(o drivers.JobOpts) error {
+	if o.VolumeBackupName == "" {
+		return fmt.Errorf("volumebackup name should be set")
+	}
+	if o.VolumeBackupNamespace == "" {
+		return fmt.Errorf("volumebackup namespace should be set")
+	}
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/drivers/kopiarestore/kopiarestore.go:120]: Added the destination pvc name check in validation during restore.

File: pkg/drivers/kopiarestore/kopiarestore.go [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+		backuplocationName,
+		"--namespace",
+		backuplocationNamespace,
+		"--repository",
+		repository,
+		"--credentials",
+		credSecretName,
+		"--target-path",
+		"/data",
+		"--id",
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/drivers/kopiarestore/kopiarestore.go:158]: done

File: pkg/executor/kopia/kopiabackup.go [hunk 14]
```diff
@@ -47,6 +47,7 @@ func newBackupCommand() *cobra.Command {
 			executor.HandleErr(runBackup(srcPath))
 		},
 	}
+	backupCommand.Flags().StringVarP(&namespace, "namespace", "n", "", "Namespace for backup command")
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/executor/kopia/kopiabackup.go:50]: There is no issue in the same flag under different sub commands. But in both the cases, this namespace is referring to backuplocation namespace so changing it to "backup-location-namespace".

File: pkg/controllers/dataexport/reconcile.go [hunk 15]
```diff
@@ -135,6 +135,37 @@ func (c *Controller) sync(ctx context.Context, in *kdmpapi.DataExport) (bool, er
...
+				vb.Spec.BackupLocation.Namespace,
+			)
+			if err != nil {
+				msg := fmt.Sprintf("failed to create cloud credential secret: %v", err)
+				logrus.Errorf(msg)
+				return false, c.updateStatus(dataExport, kdmpapi.DataExportStatusFailed, msg)
+			}
+			// For restore setting the source PVCName as the destination PVC name for the job
+			srcPVCName = dataExport.Spec.Destination.Name
+		}
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/controllers/dataexport/reconcile.go:167]: In this section, we have special handing for kopiabackup and kopiarestore drivers. IMO we don't have to add the driver checks in "DataExportStageTransferScheduled" case.

File: pkg/controllers/dataexport/reconcile.go [hunk 16]
```diff
@@ -135,6 +135,37 @@ func (c *Controller) sync(ctx context.Context, in *kdmpapi.DataExport) (bool, er
 			}
 		}
 
+		if driverName == drivers.KopiaRestore {
+			// Get the volumebackup
+			vb, err := kdmpopts.Instance().GetVolumeBackup(context.Background(),
+				dataExport.Spec.Source.Name, dataExport.Spec.Source.Namespace)
+			if err != nil {
+				if errors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/controllers/dataexport/reconcile.go:143]: removed.

File: pkg/kopia/restore.go [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+		defer b.responseLock.Unlock()
+		if err != nil {
+			b.lastError = fmt.Errorf("failed to run the restore command: %v", err)
+			logrus.Infof("stdout: %v", b.execCmd.Stdout)
+			logrus.Infof("Stderr: %v", b.execCmd.Stderr)
+			return
+		}
+
+		summaryResponse, err := getRestoreSummary(b.outBuf.Bytes(), b.errBuf.Bytes())
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/kopia/restore.go]: removed this check and put the check for variable isRunning.

File: pkg/kopia/restore.go [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+			ProgressPercentage: 100,
+			Done:               true,
+		}, nil
+	}
+
+	return &cmdexec.Status{}, nil
+}
+
+func getRestoreSummary(outBytes []byte, errBytes []byte) (*RestoreSummaryResponse, error) {
+	return &RestoreSummaryResponse{}, nil
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/kopia/restore.go]: removed this function.

File: pkg/executor/kopia/restore.go [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+				namespace,
+			); err != nil {
+				errMsg := fmt.Sprintf("failed to write a VolumeBackup status: %v", err)
+				logrus.Errorf("%v", errMsg)
+				return "", false, fmt.Errorf(errMsg)
+			}
+			return "", false, status.LastKnownError
+		}
+
+		if err = executor.WriteVolumeBackupStatus(
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/executor/kopia/restore.go:101]: done

File: pkg/kopia/restore.go (hunk 4 above)
Comment (diptiranjanpx) [reply pkg/kopia/restore.go]: removed.

File: pkg/kopia/restore.go (hunk 3 above)
Comment (diptiranjanpx) [reply pkg/kopia/restore.go]: done

File: pkg/executor/kopia/restore.go [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+	repo, err := executor.ParseCloudCred()
+	repo.Name = kopiaRepo
+
+	if err != nil {
+		if statusErr := executor.WriteVolumeBackupStatus(
+			&executor.Status{LastKnownError: err},
+			volumeBackupName,
+			namespace,
+		); statusErr != nil {
+			return statusErr
```
[the remark this answers is not in the document]
Comment (diptiranjanpx) [reply pkg/executor/kopia/restore.go]: done

**Review by prashanthpx — APPROVED** [review]

**Pull request merged**